File: android_app/src/main/java/me/timeto/app/ui/TabTimerView.kt
package me.timeto.app.ui

import androidx.annotation.DrawableRes
import androidx.compose.animation.*
import androidx.compose.animation.core.*
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Shadow
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.*
import me.timeto.app.R
import me.timeto.shared.data.TimerTabActivityData
import me.timeto.shared.vm.TabTimerVM

private val timerButtonsHeight = 28.dp
private val emojiWidth = 56.dp
private val activitiesInnerHPadding = 12.dp
private val timerHintsHPadding = 4.dp
private val menuTextButtonHPadding = (activitiesInnerHPadding + timerHintsHPadding) / 2

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun TabTimerView() {

    val (vm, state) = rememberVM { TabTimerVM() }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(c.bg),
    ) {

        Row(
            modifier = Modifier
                .padding(top = 12.dp, bottom = 10.dp, end = menuTextButtonHPadding)
                .fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
        ) {

            Icon(
                painterResource(R.drawable.sf_line_3_horizontal_small_thin),
                contentDescription = "Menu",
                tint = c.blue,
                modifier = Modifier
                    .padding(start = 12.dp)
                    .offset(y = 1.dp)
                    .clip(roundedShape)
                    .clickable {
                        Sheet.show { layer ->
                            SettingsSheet(layer = layer)
                        }
                    }
                    .size(32.dp)
                    .padding(5.dp)
            )

            SpacerW1()

            Text(
                text = "Edit",
                modifier = Modifier
                    .clip(squircleShape)
                    .clickable {
                        Sheet.show { layer ->
                            EditActivitiesSheet(layer = layer)
                        }
                    }
                    .padding(horizontal = menuTextButtonHPadding, vertical = 4.dp),
                color = c.blue,
                fontSize = 16.sp,
                fontWeight = FontWeight.Light,
            )
        }

        LazyColumn(
            contentPadding = PaddingValues(vertical = 14.dp),
        ) {

            val activitiesUI = state.activitiesUI
            itemsIndexed(
                activitiesUI,
                key = { _, i -> i.activity.id }
            ) { _, activityUI ->

                val timerData = activityUI.data.timerData
                val isActive = timerData != null
                val bgAnimate = animateColorAsState(
                    timerData?.color?.toColor() ?: c.bg,
                    spring(stiffness = Spring.StiffnessMediumLow)
                )

                SwipeToAction(
                    isStartOrEnd = remember { mutableStateOf(null) },
                    modifier = Modifier,
                    ignoreOneAction = remember { mutableStateOf(false) },
                    startView = {
                        SwipeToAction__StartView(
                            text = "Edit",
                            bgColor = c.blue
                        )
                    },
                    endView = { state ->
                        SwipeToAction__DeleteView(
                            state = state,
                            note = activityUI.deletionHint,
                            deletionConfirmationNote = activityUI.deletionConfirmation,
                        ) {
                            activityUI.delete()
                        }
                    },
                    onStart = {
                        Sheet.show { layer ->
                            ActivityFormSheet(
                                layer = layer,
                                activity = activityUI.activity,
                            )
                        }
                        false
                    },
                    onEnd = {
                        true
                    },
                    toVibrateStartEnd = listOf(true, false),
                ) {

                    Box(
                        modifier = Modifier
                            .background(bgAnimate.value)
                            .clickable {
                                Sheet.show { layer ->
                                    ActivityTimerSheet(
                                        layer = layer,
                                        activity = activityUI.activity,
                                        timerContext = null,
                                    )
                                }
                            },
                        contentAlignment = Alignment.TopCenter,
                    ) {

                        Column(
                            modifier = Modifier
                                .defaultMinSize(minHeight = 50.dp)
                                .padding(top = 8.dp, bottom = 8.dp),
                            verticalArrangement = Arrangement.Center,
                        ) {

                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(end = activitiesInnerHPadding),
                                verticalAlignment = Alignment.CenterVertically,
                            ) {

                                Text(
                                    text = activityUI.activity.emoji,
                                    style = TextStyle(
                                        shadow = Shadow(
                                            color = c.white,
                                            blurRadius = 2f,
                                        )
                                    ),
                                    modifier = Modifier
                                        .width(emojiWidth),
                                    textAlign = TextAlign.Center,
                                    fontSize = if (isActive) 20.sp else 22.sp,
                                )

                                VStack(
                                    modifier = Modifier
                                        .weight(1f),
                                ) {

                                    HStack {
                                        val textFontSize = if (isActive) 17 else 16
                                        Text(
                                            text = activityUI.data.text,
                                            modifier = Modifier
                                                .weight(1f, fill = false),
                                            color = if (isActive) c.white else c.text,
                                            fontSize = textFontSize.sp,
                                            fontWeight = if (isActive) FontWeight.Medium else FontWeight.Normal,
                                            maxLines = 1,
                                            overflow = TextOverflow.Ellipsis,
                                        )
                                        TriggersListIconsView(activityUI.data.textTriggers, (textFontSize - 2).sp)
                                    }

                                    val listNote = activityUI.data.note
                                    if (listNote != null)
                                        HStack(
                                            modifier = Modifier
                                                .offset(y = (-1).dp),
                                            verticalAlignment = Alignment.CenterVertically,
                                        ) {
                                            val noteIcon = activityUI.data.noteIcon
                                            if (noteIcon != null) {
                                                val noteIconResId: Int = when (noteIcon) {
                                                    TimerTabActivityData.NoteIcon.event -> R.drawable.sf_calendar_medium_light
                                                }
                                                Icon(
                                                    painterResource(noteIconResId),
                                                    contentDescription = "Event",
                                                    tint = c.white,
                                                    modifier = Modifier
                                                        .padding(end = 5.dp)
                                                        .size(14.dp),
                                                )
                                            }
                                            Text(
                                                text = listNote,
                                                modifier = Modifier
                                                    .weight(1f, fill = false),
                                                color = c.white,
                                                fontSize = 14.sp,
                                                fontWeight = FontWeight.Light,
                                            )
                                            TriggersListIconsView(activityUI.data.noteTriggers, 12.sp)
                                        }
                                }

                                activityUI.timerHints.forEach { hintUI ->
                                    Text(
                                        text = hintUI.text,
                                        modifier = Modifier
                                            .padding(top = 1.dp)
                                            .clip(roundedShape)
                                            .clickable {
                                                hintUI.startInterval()
                                            }
                                            .padding(horizontal = timerHintsHPadding, vertical = 3.dp),
                                        color = if (isActive) c.white else c.blue,
                                        fontSize = 14.sp,
                                        fontWeight = FontWeight.Light,
                                    )
                                }
                            }

                            if (timerData != null) {

                                ZStack(
                                    modifier = Modifier
                                        .padding(top = 8.dp, bottom = 2.dp, start = 12.dp, end = 10.dp)
                                        .fillMaxWidth(),
                                ) {

                                    val timerDataTitleLen = timerData.title.length
                                    val timerTitleFontSize = when {
                                        timerDataTitleLen <= 5 -> 34.sp
                                        timerDataTitleLen <= 7 -> 32.sp
                                        else -> 28.sp
                                    }

                                    Text(
                                        text = timerData.title,
                                        modifier = Modifier
                                            .clickable {
                                                vm.toggleIsPurple()
                                            }
                                            .align(Alignment.BottomStart),
                                        fontFamily = timerFont,
                                        fontSize = timerTitleFontSize,
                                        color = c.white,
                                    )

                                    HStack(
                                        modifier = Modifier
                                            .padding(top = 8.dp)
                                            .align(Alignment.BottomEnd),
                                    ) {

                                        Icon(
                                            painterResource(R.drawable.sf_pause_small_medium),
                                            contentDescription = "Pause",
                                            tint = c.white,
                                            modifier = Modifier
                                                .size(timerButtonsHeight)
                                                .border(1.dp, c.white, roundedShape)
                                                .clip(roundedShape)
                                                .clickable {
                                                    activityUI.pauseLastInterval()
                                                }
                                                .padding(8.dp),
                                        )

                                        HStack(
                                            verticalAlignment = Alignment.CenterVertically,
                                            modifier = Modifier
                                                .padding(start = 8.dp)
                                                .height(timerButtonsHeight)
                                                .border(1.dp, c.white, roundedShape)
                                                .clip(roundedShape)
                                                .clickable {
                                                    timerData.restart()
                                                }
                                                .padding(start = 7.dp, end = 6.dp),
                                        ) {

                                            Icon(
                                                painterResource(id = R.drawable.sf_clock_arrow_circlepath_small_light),
                                                contentDescription = "Restart",
                                                tint = c.white,
                                                modifier = Modifier
                                                    .size(15.dp),
                                            )

                                            Text(
                                                text = timerData.restartText,
                                                modifier = Modifier
                                                    .padding(start = 3.dp, bottom = 1.dp),
                                                fontSize = 14.sp,
                                                fontWeight = FontWeight.Light,
                                                color = c.white,
                                            )
                                        }
                                    }
                                }
                            }
                        }

                        DividerBg(
                            modifier = Modifier.padding(start = emojiWidth),
                            isVisible = activityUI.withTopDivider,
                        )
                    }
                }
            }

            item {

                HStack(
                    modifier = Modifier
                        .padding(top = 12.dp, start = 10.dp),
                ) {

                    ChartHistoryButton(
                        text = "Chart",
                        iconResId = R.drawable.sf_chart_pie_small_thin,
                    ) {
                        Dialog.show(
                            modifier = Modifier.fillMaxHeight(0.95f),
                        ) { layer ->
                            ChartDialogView(layer::close)
                        }
                    }

                    ChartHistoryButton(
                        "History",
                        iconResId = R.drawable.sf_list_bullet_rectangle_small_thin,
                        iconSize = 20.dp,
                    ) {
                        Dialog.show(
                            modifier = Modifier.fillMaxHeight(0.95f),
                        ) { layer ->
                            HistoryDialogView(layer::close)
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun ChartHistoryButton(
    text: String,
    @DrawableRes iconResId: Int,
    iconSize: Dp = 18.dp,
    onClick: () -> Unit,
) {
    HStack(
        modifier = Modifier
            .clip(squircleShape)
            .clickable {
                onClick()
            }
            .padding(horizontal = menuTextButtonHPadding, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {

        Icon(
            painterResource(iconResId),
            contentDescription = text,
            tint = c.blue,
            modifier = Modifier
                .padding(end = 5.dp)
                .size(iconSize)
        )

        Text(
            text = text,
            color = c.blue,
            fontSize = 16.sp,
            fontWeight = FontWeight.Light,
        )
    }
}

File: ios_app/iphone/TabTimerView.swift
import SwiftUI
import shared

private let emojiWidth = 58.0
private let activitiesInnerHPadding = 12.0
private let timerHintsHPadding = 4.0

struct TabTimerView: View {

    @State private var vm = TabTimerVM()

    @State private var isReadmePresented = false

    @State private var isEditActivitiesPresented = false
    @State private var isSettingsSheetPresented = false

    @State private var isChartPresented = false
    @State private var isHistoryPresented = false

    var body: some View {

        // Without NavigationView the NavigationLink does not work,
        // remember the .navigationBarHidden(true)
        NavigationView {

            // If outside of the NavigationView - deletion swipe does not opened in full size.
            VMView(vm: vm) { state in

                ZStack {

                    Color(.bg).edgesIgnoringSafeArea(.all)

                    //
                    // Top Menu Bar

                    VStack {

                        HStack {

                            Button(
                                    action: {
                                        isSettingsSheetPresented.toggle()
                                    },
                                    label: {
                                        Image(systemName: "line.3.horizontal")
                                                .foregroundColor(.blue)
                                                .font(.system(size: 23, weight: .thin))
                                    }
                            )
                                    .padding(.leading, 15 - onePx)

                            Spacer()

                            Button(
                                    action: { isEditActivitiesPresented = true },
                                    label: {
                                        Text("Edit")
                                                .font(.system(size: 17, weight: .light))
                                    }
                            )
                                    .padding(.trailing, activitiesInnerHPadding + timerHintsHPadding)
                        }
                                .padding(.top, 12)
                                .padding(.bottom, 10)

                        //
                        // List

                        ScrollView(.vertical, showsIndicators: false) {

                            VStack {

                                MySpacerSize(height: 20)

                                VStack {

                                    let activitiesUI = state.activitiesUI

                                    ForEach(activitiesUI, id: \.activity.id) { activityUI in
                                        ActivityRowView(
                                                vm: vm,
                                                activityUI: activityUI,
                                                lastInterval: state.lastInterval,
                                                withTopDivider: activityUI.withTopDivider
                                        )
                                    }
                                }
                                /*
                                    .onDelete { set in
                                        // to show the button
                                    }
                                    .onMove { set, toIdx in
                                        if (set.count != 1) {
                                            fatalError("bad count for moving \(set.count)")
                                        }

                                        var tempList = activities.map {
                                            $0
                                        }

                                        let fromIdx = set.first!
                                        if fromIdx < toIdx {
                                            tempList.insert(tempList[fromIdx], at: toIdx)
                                            tempList.remove(at: fromIdx)
                                        } else {
                                            let removedTask = tempList.remove(at: fromIdx)
                                            tempList.insert(removedTask, at: toIdx)
                                        }

                                        for (tempIndex, tempTask) in tempList.enumerated() {
                                            tempTask.updateSort(tempIndex)
                                        }
                                    }
                                     */
                            }

                            HStack {

                                ChartHistoryButton(text: "Chart", iconName: "chart.pie", iconSize: 18) {
                                    isChartPresented = true
                                }
                                        .padding(.leading, 18)
                                        .padding(.trailing, activitiesInnerHPadding + timerHintsHPadding)
                                        .sheetEnv(isPresented: $isChartPresented) {
                                            VStack {

                                                ChartView()
                                                        .padding(.top, 15)

                                                Button(
                                                        action: { isChartPresented.toggle() },
                                                        label: { Text("close").fontWeight(.light) }
                                                )
                                                        .padding(.bottom, 4)
                                            }
                                        }

                                ChartHistoryButton(text: "History", iconName: "list.bullet.rectangle", iconSize: 18) {
                                    isHistoryPresented = true
                                }
                                        .sheetEnv(isPresented: $isHistoryPresented) {
                                            ZStack {
                                                Color(.myBackground).edgesIgnoringSafeArea(.all)
                                                HistoryView(isHistoryPresented: $isHistoryPresented)
                                            }
                                                    // todo
                                                    .interactiveDismissDisabled()
                                        }

                                Spacer()
                            }
                                    .padding(.top, 8)

                            MySpacerSize(height: 24)
                        }
                    }
                }
                        .sheetEnv(
                                isPresented: $isEditActivitiesPresented
                        ) {
                            EditActivitiesSheet(
                                    isPresented: $isEditActivitiesPresented
                            )
                        }
                        .sheetEnv(
                                isPresented: $isSettingsSheetPresented
                        ) {
                            SettingsSheet(isPresented: $isSettingsSheetPresented)
                        }
                        .sheetEnv(isPresented: $isReadmePresented) {
                            TabReadmeView(isPresented: $isReadmePresented)
                        }
                        .navigationBarHidden(true)
            }
        }
    }
}

private struct ChartHistoryButton: View {

    let text: String
    let iconName: String
    let iconSize: CGFloat
    let onClick: () -> Void

    var body: some View {
        Button(
                action: { onClick() },
                label: {
                    HStack {
                        Image(systemName: iconName)
                                .font(.system(size: iconSize, weight: .thin))
                                .padding(.trailing, 3 + onePx)
                        Text(text)
                                .font(.system(size: 17, weight: .light))
                    }
                }
        )
    }
}

private struct ActivityRowView: View {

    var vm: TabTimerVM
    var activityUI: TabTimerVM.ActivityUI
    var lastInterval: IntervalModel
    var withTopDivider: Bool

    @State private var isSetTimerPresented = false
    @State private var isEditSheetPresented = false

    @State private var isActiveAnim = false
    @State private var bgColorAnim = Color(.bg)

    @EnvironmentObject private var timetoSheet: TimetoSheet

    var body: some View {
        MyListSwipeToActionItem(
                deletionHint: activityUI.deletionHint,
                deletionConfirmationNote: activityUI.deletionConfirmation,
                onEdit: {
                    isEditSheetPresented = true
                },
                onDelete: {
                    activityUI.delete()
                }
        ) {
            ZStack(alignment: .top) {
                AnyView(safeView)
                if withTopDivider {
                    DividerBg()
                            .padding(.leading, emojiWidth)
                }
            }
                    // todo remove after removing MyListSwipeToActionItem()
                    .background(Color(.bg))
        }
                .animateVmValue(value: activityUI.data.timerData != nil, state: $isActiveAnim)
                .animateVmValue(value: activityUI.data.timerData?.color.toColor() ?? Color(.bg), state: $bgColorAnim)
    }

    private var safeView: some View {

        Button(
                action: {
                    timetoSheet.showActivityTimerSheet(
                            activity: activityUI.activity,
                            isPresented: $isSetTimerPresented,
                            timerContext: nil,
                            onStart: {
                                isSetTimerPresented.toggle()
                                /// With animation twitching emoji
                            }
                    )
                },
                label: {

                    VStack(alignment: .leading) {

                        HStack {

                            Text(activityUI.activity.emoji)
                                    .frame(width: emojiWidth)
                                    .font(.system(size: 25))
                                    .animation(nil, value: isActiveAnim)
                                    .shadow(color: .white.opacity(0.4), radius: 1)

                            VStack {

                                HStack {

                                    let textFontSize = isActiveAnim ? 18.0 : 17.0

                                    HStack {
                                        Text(activityUI.data.text)
                                                .font(.system(size: textFontSize, weight: isActiveAnim ? .semibold : .regular))
                                                .animation(nil, value: isActiveAnim)
                                                .foregroundColor(isActiveAnim ? .white : Color(.label))
                                                .truncationMode(.tail)
                                                .lineLimit(1)
                                        TriggersListIconsView(triggers: activityUI.data.textTriggers, fontSize: textFontSize - 2)
                                    }
                                }
                                        .frame(maxWidth: .infinity, alignment: .leading)

                                if let note = activityUI.data.note {
                                    HStack {
                                        if let noteIcon = activityUI.data.noteIcon {
                                            let systemName: String = {
                                                if noteIcon == .event { return "calendar" }
                                                fatalError()
                                            }()
                                            Image(systemName: systemName)
                                                    .foregroundColor(.white)
                                                    .font(.system(size: 14, weight: .regular))
                                                    .padding(.trailing, 4)
                                        }
                                        Text(note)
                                                .foregroundColor(.white)
                                                .font(.system(size: 15, weight: .light))
                                        TriggersListIconsView(triggers: activityUI.data.noteTriggers, fontSize: 13)
                                    }
                                            .frame(maxWidth: .infinity, alignment: .leading)
                                }
                            }

                            Spacer()

                            ForEach(activityUI.timerHints, id: \.self) { hintUI in
                                Button(
                                        action: {
                                            hintUI.startInterval {}
                                        },
                                        label: {
                                            Text(hintUI.text)
                                                    .offset(y: onePx)
                                                    .font(.system(size: 15, weight: .light))
                                                    .foregroundColor(isActiveAnim ? .white : .blue)
                                                    .padding(.horizontal, timerHintsHPadding)
                                        }
                                )
                            }
                        }
                                .padding(.trailing, activitiesInnerHPadding)

                        if let timerData = activityUI.data.timerData {

                            ZStack(alignment: .bottomLeading) {

                                let timerDataTitleLen = timerData.title.count
                                let timerTitleFontWeight: CGFloat = {
                                    if timerDataTitleLen <= 5 { return 38 }
                                    if timerDataTitleLen <= 7 { return 35 }
                                    return 30
                                }()

                                Button(
                                        action: {
                                            vm.toggleIsPurple()
                                        },
                                        label: {
                                            Text(timerData.title)
                                                    .padding(.leading, activitiesInnerHPadding - 1)
                                                    .font(getTimerFont(size: timerTitleFontWeight))
                                                    .foregroundColor(.white)
                                        }
                                )

                                HStack {

                                    Spacer()

                                    HStack {

                                        let timerButtonsHeight = 30.0

                                        Button(
                                                action: {
                                                    activityUI.pauseLastInterval()
                                                },
                                                label: {
                                                    Image(systemName: "pause")
                                                            .foregroundColor(.white)
                                                            .font(.system(size: 14, weight: .regular))
                                                            .frame(width: timerButtonsHeight, height: timerButtonsHeight)
                                                            .overlay(roundedShape.stroke(Color.white, lineWidth: 1))
                                                }
                                        )

                                        Button(
                                                action: {
                                                    timerData.restart()
                                                },
                                                label: {
                                                    HStack {

                                                        Image(systemName: "clock.arrow.circlepath")
                                                                .foregroundColor(.white)
                                                                .font(.system(size: 16, weight: .light))

                                                        Text(timerData.restartText)
                                                                .padding(.leading, 2)
                                                                .padding(.trailing, 2)
                                                                .font(.system(size: 15, weight: .light))
                                                                .foregroundColor(.white)
                                                    }
                                                            .padding(.horizontal, 6)
                                                            .frame(height: timerButtonsHeight)
                                                            .overlay(roundedShape.stroke(Color.white, lineWidth: 1))
                                                }
                                        )
                                                .padding(.leading, 8)
                                                .padding(.trailing, activitiesInnerHPadding - 2)
                                    }
                                            .padding(.bottom, 1)
                                }
                            }
                                    .padding(.top, 8)
                                    .padding(.bottom, 1)
                        }
                    }
                            .padding(.top, 10)
                            .padding(.bottom, 10)
                            /// #TruncationDynamic + README_APP.md
                            .id("\(activityUI.activity.id) \(lastInterval.note)")
                }
        )
                .sheetEnv(isPresented: $isEditSheetPresented) {
                    ActivityFormSheet(
                            isPresented: $isEditSheetPresented,
                            editedActivity: activityUI.activity
                    ) {
                    }
                }
                .buttonStyle(ActivityButtonStyle(bgColor: bgColorAnim, isActive: isActiveAnim))
    }
}

private struct ActivityButtonStyle: ButtonStyle {

    let bgColor: Color
    let isActive: Bool

    func makeBody(configuration: Self.Configuration) -> some View {
        configuration.label.background(calcColor(configuration: configuration))
    }

    private func calcColor(configuration: Self.Configuration) -> Color {
        if !configuration.isPressed { return bgColor }
        if isActive { return bgColor.opacity(0.8) }
        return Color(.systemGray5)
    }
}
